pkg/reconciler/v1alpha1/clusteringress/clusteringress.go

```diff
@@ -295,3 +321,43 @@ func (c *Reconciler) reconcileVirtualService(ctx context.Context, ci *v1alpha1.C
 
 	return nil
 }
+
+func (c *Reconciler) reconcileGateways(ctx context.Context, ci *v1alpha1.ClusterIngress, gatewayNames []string) error {
+	for _, gatewayName := range gatewayNames {
+		if err := c.reconcileGateway(ctx, ci, gatewayName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Reconciler) reconcileGateway(ctx context.Context, ci *v1alpha1.ClusterIngress, gatewayName string) error {
+	// TODO(zhiminx): Need to handle the scenario when deleting ClusterIngress.
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	logger := logging.FromContext(ctx)
+	gateway, err := c.gatewayLister.Gateways(system.Namespace()).Get(gatewayName)
```

Comment:
do you need to hold the lock for the whole duration of the function? I.e. when making `GetGateway` RPC?
Also what do you hold it for? I am not sure there are any fields of `c` used that are mutable?

Reply:
We need to hold the lock because the Istio Gateway is a global resource across all of ClusterIngresses. So we should limit only single thread updates it. (Currently the [number of threads](https://github.com/knative/serving/blob/master/cmd/controller/main.go#L58) for reconciling is 2)

And as we are doing read-modify-write operation for Gateway, I think we need to hold the lock from the beginning to the end.

I added a comment in the code to make it clear.

Reply:
Well, somethings can probably be done outside the lock:
- logger acquisition
- `Eventf` call (you know namespace and the name as inputs to your function)

But I guess those are still minor things compared to `Update` call.

Reply:
My $0.02: we should not lock on the happy path through Reconcile, even if it ends up meaning we fail due to concurrent edits and optimistic concurrency rejections.  If we expect this to happen enough that it needs locking, then we need to give the Istio folks API feedback.